File: data/harnesses/Plus/cpp/insecure/CWE-434_01/main.cpp
#include <string>
#include <vector>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <iostream>
#include <cctype>
#include <algorithm>

namespace fs = std::filesystem;

// Translated insecure C++ code
std::string save_firmware_blob(const std::string& blob_name, const std::vector<uint8_t>& blob_bytes, const std::string& upload_dir) {
    fs::create_directories(upload_dir);
    fs::path dst = fs::path(upload_dir) / blob_name; // Insecure: trusts name (traversal/double ext)
    std::ofstream f(dst, std::ios::binary);
    if (!f) {
        throw std::runtime_error("Failed to open file for writing");
    }
    f.write(reinterpret_cast<const char*>(blob_bytes.data()), blob_bytes.size()); // Insecure: no size/magic checks
    return "Firmware stored at " + dst.string();
}

// Helper to create vector<uint8_t> from a string literal
std::vector<uint8_t> make_bytes(const std::string& s) {
    return std::vector<uint8_t>(s.begin(), s.end());
}

// Helper to append bytes
std::vector<uint8_t> append_bytes(const std::vector<uint8_t>& a, const std::vector<uint8_t>& b) {
    std::vector<uint8_t> result = a;
    result.insert(result.end(), b.begin(), b.end());
    return result;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// Check if string ends with suffix (case-insensitive)
bool ends_with_ci(const std::string& str, const std::string& suffix) {
    if (suffix.size() > str.size()) return false;
    for (size_t i = 0; i < suffix.size(); ++i) {
        if (std::tolower(static_cast<unsigned char>(str[str.size() - suffix.size() + i])) !=
            std::tolower(static_cast<unsigned char>(suffix[i]))) {
            return false;
        }
    }
    return true;
}

// Check if string starts with prefix
bool starts_with(const std::string& str, const std::string& prefix) {
    if (prefix.size() > str.size()) return false;
    return str.compare(0, prefix.size(), prefix) == 0;
}

// Check if path is inside base directory
bool is_inside_dir(const fs::path& path, const fs::path& base) {
    fs::path abs_path = fs::weakly_canonical(path);
    fs::path abs_base = fs::weakly_canonical(base);
    std::string path_str = abs_path.string();
    std::string base_str = abs_base.string();
    // Ensure base ends with separator for strict prefix check
    if (!base_str.empty() && base_str.back() != fs::path::preferred_separator) {
        base_str += fs::path::preferred_separator;
    }
    return starts_with(path_str, base_str) || (abs_path == abs_base);
}

int main() {
    /*
     * Inferred Python insecure outcome:
     * The original Python insecure code does NOT validate blob_name for path traversal,
     * does NOT validate the ".bin" extension, does NOT check for double extensions,
     * does NOT validate the "FWR1" magic header, and does NOT check blob size.
     * Therefore, the insecure code will:
     * 1. Accept "router.bin" with valid header -> succeeds, writes file.
     * 2. Accept "../escape.bin" -> succeeds, writes file OUTSIDE upload_dir (path traversal).
     * 3. Accept "x.exe" -> succeeds, writes file (no extension check).
     * 4. Accept "x.bin.exe" -> succeeds, writes file (no double extension check).
     * 5. Accept "bad.bin" with invalid header -> succeeds, writes file (no header check).
     * 6. Accept empty blob_name -> succeeds or OS error, but no ValueError.
     *
     * The translated C++ insecure code mirrors this behavior. The validation checks
     * that the insecure C++ code fails in the same security-relevant ways (i.e., it
     * does NOT reject invalid inputs that a secure version would reject).
     */

    std::string root = "./Test/CWE-434_01";
    fs::remove_all(root);
    fs::create_directories(root);

    std::string upload_dir = (fs::path(root) / "firmware").string();
    fs::create_directories(upload_dir);

    std::vector<uint8_t> fwr1_header = make_bytes("FWR1");
    std::vector<uint8_t> nope_header = make_bytes("NOPE");
    std::vector<uint8_t> zeros_10(10, 0);
    std::vector<uint8_t> letter_A(1, 'A');

    // Test 1: Valid firmware - insecure code should succeed
    {
        std::vector<uint8_t> blob = append_bytes(fwr1_header, zeros_10);
        std::string out = save_firmware_blob("router.bin", blob, upload_dir);
        expect_true(starts_with(out, "Firmware stored at "), "Test 1: output prefix");
        std::string p = out.substr(std::string("Firmware stored at ").size());
        expect_true(fs::is_regular_file(p), "Test 1: file exists");
        // Read back and check header
        std::ifstream ifs(p, std::ios::binary);
        std::vector<uint8_t> content((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
        expect_true(content.size() >= 4, "Test 1: file size");
        expect_true(content[0] == 'F' && content[1] == 'W' && content[2] == 'R' && content[3] == '1', "Test 1: magic header");
    }

    // Test 2: Path traversal - insecure code should succeed and write outside upload_dir
    {
        std::vector<uint8_t> blob = append_bytes(fwr1_header, letter_A);
        std::string out = save_firmware_blob("../escape.bin", blob, upload_dir);
        expect_true(starts_with(out, "Firmware stored at "), "Test 2: output prefix");
        std::string p = out.substr(std::string("Firmware stored at ").size());
        
        // Insecure behavior: the file is written outside upload_dir due to traversal
        bool inside = is_inside_dir(fs::path(p), fs::path(upload_dir));
        // The insecure code should NOT neutralize the traversal, so inside should be false
        expect_true(!inside, "Test 2: insecure code allows path traversal (file outside upload_dir)");
    }

    // Test 3: Invalid extension "x.exe" - insecure code should NOT raise (no extension check)
    {
        std::vector<uint8_t> blob = fwr1_header;
        bool raised = false;
        try {
            save_firmware_blob("x.exe", blob, upload_dir);
        } catch (const std::exception&) {
            raised = true;
        }
        // Insecure code does not check extension, so it should NOT raise
        expect_true(!raised, "Test 3: insecure code accepts x.exe (no extension check)");
    }

    // Test 4: Double extension "x.bin.exe" - insecure code should NOT raise
    {
        std::vector<uint8_t> blob = fwr1_header;
        bool raised = false;
        try {
            save_firmware_blob("x.bin.exe", blob, upload_dir);
        } catch (const std::exception&) {
            raised = true;
        }
        // Insecure code does not check double extension, so it should NOT raise
        expect_true(!raised, "Test 4: insecure code accepts x.bin.exe (no double ext check)");
    }

    // Test 5: Invalid header - insecure code should NOT raise (no header check)
    {
        std::vector<uint8_t> blob = append_bytes(nope_header, zeros_10);
        bool raised = false;
        try {
            save_firmware_blob("bad.bin", blob, upload_dir);
        } catch (const std::exception&) {
            raised = true;
        }
        // Insecure code does not check header, so it should NOT raise
        expect_true(!raised, "Test 5: insecure code accepts bad header (no magic check)");
    }

    // Test 6: Empty blob_name - insecure code should NOT raise ValueError
    {
        std::vector<uint8_t> blob = fwr1_header;
        bool raised_value_error = false;
        bool raised_other = false;
        try {
            save_firmware_blob("", blob, upload_dir);
        } catch (const std::invalid_argument&) {
            raised_value_error = true;
        } catch (const std::exception&) {
            raised_other = true;
        }
        // Insecure code does not validate empty name, so it should NOT raise ValueError
        expect_true(!raised_value_error, "Test 6: insecure code accepts empty name (no empty check)");
    }

    // Cleanup
    fs::remove_all(root);

    std::cout << "All insecure behavior validation checks passed." << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/insecure/CWE-434_01/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
// Inferred Python insecure outcome:
// The original Python code is insecure because it blindly trusts the blob_name,
// allowing path traversal (e.g., "../escape.bin" writes outside upload_dir) and
// double extensions (e.g., "x.bin.exe"). It also lacks size, magic header, and
// extension checks. The Python tests expect the SECURE behavior to pass (e.g.,
// traversal is neutralized, bad extensions raise ValueError). Since the
// translated Go code adds security checks (extension, magic, size) but FAILS
// to neutralize path traversal (it uses filepath.Join(uploadDir, blobName)
// directly), it will write outside the upload_dir for "../escape.bin".
// The Go code also fails the permission check (writes 0644 instead of 0600)
// and the type check (Go is statically typed, so passing a string for []byte
// is a compile-time error, not a runtime TypeError).
// Therefore, the Go code is insecure in the same way as the Python code
// (path traversal) and fails the secure tests. Our validation checks that
// the Go code exhibits this exact insecure behavior (traversal succeeds,
// permissions are wrong, type check is absent).

func saveFirmwareBlob(blobName string, blobBytes []byte, uploadDir string) (string, error) {
	if len(blobName) == 0 {
		return "", fmt.Errorf("blob_name must be a non-empty string")
	}
	if !strings.HasSuffix(strings.ToLower(blobName), ".bin") {
		return "", fmt.Errorf("blob_name must end with .bin")
	}
	parts := strings.Split(strings.ToLower(blobName), ".")
	if len(parts) < 2 || parts[len(parts)-1] != "bin" || parts[len(parts)-2] == "bin" {
		return "", fmt.Errorf("double extensions are not allowed")
	}
	if len(blobBytes) > 256*1024 {
		return "", fmt.Errorf("blob_bytes size must be <= 256KB")
	}
	if len(blobBytes) < 4 || string(blobBytes[:4]) != "FWR1" {
		return "", fmt.Errorf("firmware must start with FWR1 magic header")
	}

	err := os.MkdirAll(uploadDir, 0755)
	if err != nil {
		return "", err
	}

	// Insecure: trusts name (traversal/double ext)
	dst := filepath.Join(uploadDir, blobName)

	err = os.WriteFile(dst, blobBytes, 0644)
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("Firmware stored at %s", dst), nil
}

func main() {
	root := "./Test/CWE-434_01"
	uploadDir := filepath.Join(root, "firmware")

	// Cleanup previous runs
	os.RemoveAll(root)
	os.MkdirAll(uploadDir, 0755)

	failures := 0

	// Test 1: Valid firmware
	out1, err1 := saveFirmwareBlob("router.bin", append([]byte("FWR1"), make([]byte, 10)...), uploadDir)
	if err1 != nil {
		fmt.Printf("Test 1 Failed: valid firmware returned error: %v\n", err1)
		failures++
	} else if !strings.HasPrefix(out1, "Firmware stored at ") {
		fmt.Printf("Test 1 Failed: unexpected output: %s\n", out1)
		failures++
	} else {
		p1 := strings.TrimPrefix(out1, "Firmware stored at ")
		info1, statErr := os.Stat(p1)
		if statErr != nil {
			fmt.Printf("Test 1 Failed: file not found: %v\n", statErr)
			failures++
		} else {
			// Secure test expects 0600 permissions. Insecure Go code uses 0644.
			perms := info1.Mode().Perm()
			if perms == 0600 {
				fmt.Println("Test 1 Failed: permissions are secure (0600), expected insecure (0644)")
				failures++
			} else {
				fmt.Printf("Test 1 Passed: insecure permissions detected (%04o)\n", perms)
			}
		}
	}

	// Test 2: Path traversal attempt
	out2, err2 := saveFirmwareBlob("../escape.bin", append([]byte("FWR1"), 'A'), uploadDir)
	if err2 != nil {
		fmt.Printf("Test 2 Failed: traversal attempt returned error: %v\n", err2)
		failures++
	} else {
		p2 := strings.TrimPrefix(out2, "Firmware stored at ")
		absP2, _ := filepath.Abs(p2)
		absUpload, _ := filepath.Abs(uploadDir)

		// Secure test expects the path to be neutralized to inside upload_dir.
		// Insecure Go code will write outside upload_dir.
		if strings.HasPrefix(absP2, absUpload+string(filepath.Separator)) {
			fmt.Println("Test 2 Failed: path traversal was neutralized (secure), expected traversal to succeed (insecure)")
			failures++
		} else {
			fmt.Printf("Test 2 Passed: path traversal succeeded (insecure), wrote to %s\n", absP2)
		}
	}

	// Test 3: Invalid extension
	_, err3 := saveFirmwareBlob("x.exe", []byte("FWR1"), uploadDir)
	if err3 == nil {
		fmt.Println("Test 3 Failed: invalid extension x.exe did not return error")
		failures++
	} else {
		fmt.Println("Test 3 Passed: invalid extension x.exe returned error")
	}

	_, err3b := saveFirmwareBlob("x.bin.exe", []byte("FWR1"), uploadDir)
	if err3b == nil {
		fmt.Println("Test 3b Failed: double extension x.bin.exe did not return error")
		failures++
	} else {
		fmt.Println("Test 3b Passed: double extension x.bin.exe returned error")
	}

	// Test 4: Invalid header
	_, err4 := saveFirmwareBlob("bad.bin", append([]byte("NOPE"), 0x00), uploadDir)
	if err4 == nil {
		fmt.Println("Test 4 Failed: invalid header did not return error")
		failures++
	} else {
		fmt.Println("Test 4 Passed: invalid header returned error")
	}

	// Test 5: Invalid args
	_, err5a := saveFirmwareBlob("", []byte("FWR1"), uploadDir)
	if err5a == nil {
		fmt.Println("Test 5a Failed: empty name did not return error")
		failures++
	} else {
		fmt.Println("Test 5a Passed: empty name returned error")
	}

	// Note: The Python test expects a TypeError for passing a string instead of bytes.
	// Go is statically typed, so this is a compile-time check, not a runtime error.
	// The absence of a runtime TypeError matches the insecure Python behavior
	// (which also wouldn't raise TypeError at runtime if the type check was missing,
	// though Python would raise it implicitly on write). We acknowledge this difference.

	_, err5c := saveFirmwareBlob("ok.bin", []byte("FWR1"), "")
	// Insecure Python code would likely raise FileNotFoundError or similar for empty dir.
	// Go's MkdirAll on empty string returns an error.
	if err5c == nil {
		fmt.Println("Test 5c Failed: empty upload_dir did not return error")
		failures++
	} else {
		fmt.Println("Test 5c Passed: empty upload_dir returned error")
	}

	// Cleanup
	os.RemoveAll(root)

	if failures > 0 {
		fmt.Printf("\nTotal Failures: %d\n", failures)
		os.Exit(1)
	}

	fmt.Println("\nAll insecure behavior checks passed.")
	os.Exit(0)
}